csrc/lower_thread_predicate.cpp

```diff
@@ -494,6 +519,185 @@ class RedundantUseAnalysis : BackwardVisitor {
+            std::find(root_domain.begin(), root_domain.end(), merge->outer());
+        auto inner_iter =
+            std::find(root_domain.begin(), root_domain.end(), merge->inner());
+        if (inner_iter != root_domain.end()) {
+          merged_root_domains.emplace_back(*inner_iter);
+          index_root_domain.emplace_back(inner_iter - root_domain.begin());
+        }
+        if (outer_iter != root_domain.end()) {
+          merged_root_domains.emplace_back(*outer_iter);
+          index_root_domain.emplace_back(outer_iter - root_domain.begin());
+        }
+      }
+    }
+    // The following sort is added because in NVFuserTest.FusionIssue2076_CUDA
+    // the order is [B2, I1, I3] while the correct order should be [I3, B2, I1]
+    int n_elements = merged_root_domains.size();
+    std::vector<int> indices(n_elements);
+    std::iota(indices.begin(), indices.end(), 0);
+    std::sort(indices.begin(), indices.end(), [&](int a, int b) {
+      return index_root_domain[a] > index_root_domain[b];
+    });
+    std::vector<IterDomain*> merged_root_domains_sorted(n_elements);
+    for (int i = 0; i < n_elements; ++i) {
+      merged_root_domains_sorted[i] = merged_root_domains[indices[i]];
+    }
+
+    return merged_root_domains_sorted;
+  };
+
+  // Set stride when use extent of merged root domains to index leaf domain.
+  // e.g. Root: [I1,B2,B3] -> Leaf: [I1B2B3], the merged_root_domains =
+  // {B3,B2,I1}. root_stride = {1, len(B3), len(B3) * len(B2)}. For broadcast
+  // root domain, use the extent of its concretized domain.
+  auto getRootStride = [&](const std::vector<IterDomain*>& merged_root_domains,
+                           const std::unordered_map<IterDomain*, IterDomain*>&
+                               concretized_broadcast_root_domains) {
+    std::vector<Val*> root_stride(merged_root_domains.size());
+    root_stride[0] = IrBuilder::create<Int>(1);
+    for (int i = 1; i < (int)merged_root_domains.size(); i++) {
+      auto pre_crd = merged_root_domains[i - 1];
+      Val* pre_extent = pre_crd->isBroadcast()
+          ? concretized_broadcast_root_domains.at(pre_crd)->extent()
+          : pre_crd->extent();
+      root_stride[i] = IrBuilder::mulExpr(root_stride[i - 1], pre_extent);
+    }
+    return root_stride;
+  };
+
+  // set write stride
+  // e.g. Root: [I1,B2,B3] -> Leaf: [I1B2B3], the merged_root_domains =
+  // {B3,B2,I1}. root_stride = {1, len(B3), len(B3) * len(B2)}. write_stride_mod
+  // = {len(B3) * len(B2)}. More cases can be found in
+  // NVFuserTest.FusionAvoidRedundantWrite_CUDA
+  auto setWriteStride = [&](const std::vector<IterDomain*>& merged_root_domains,
+                            const std::vector<Val*>& root_stride,
+                            const ParallelType& pt) {
+    int ndim = (int)merged_root_domains.size();
+    if (ndim > 3) {
+      // only tested up to 3 dims merged into 1 dim
+      return;
+    }
+
+    // backward iterator, to ensure the visit is from outer to inner dims
```

Comment:
Is this needed because your merged_root_domain is reversed order from the root ordering?

Reply:
yes.

Reply:
Does it need to be written in reverse order?

Reply:
reverse order is convinent for the calculation of strides. Of course, we can also store it in unreversed order and visit backward when calculate stride.

Reply:
I kind of like the latter approach, as it keeps what's stored in a 1:1 "zippable" order. Generally changing order of iterating on a data structure makes a lot of sense to me, flipping the order of a data structure relative to how it's generally been stored is a bit more "mentally complex" in my opinion and should be done if storing as such improves cache locallity/performance (i.e. when we change layout of tensors to improve locality of tiles as they're accessed).